File: cmd/arkfile-client/output_basename.go
package main

import (
	"fmt"
	"path/filepath"
	"strings"
)

func splitBasenameExtension(filename string) (stem, ext string) {
	base := filepath.Base(filename)
	if base == "." || base == "/" || base == "" {
		base = "download"
	}
	// filepath.Ext includes the dot; treat leading-dot names as stem-only.
	ext = filepath.Ext(base)
	if ext == "" || ext == base {
		return base, ""
	}
	stem = strings.TrimSuffix(base, ext)
	if stem == "" {
		return base, ""
	}
	return stem, ext
}

func nextAvailableBasename(desiredName string, taken map[string]struct{}) string {
	stem, ext := splitBasenameExtension(desiredName)
	candidate := stem + ext
	if _, exists := taken[candidate]; !exists {
		return candidate
	}
	for n := 1; ; n++ {
		candidate = fmt.Sprintf("%s-%d%s", stem, n, ext)
		if _, exists := taken[candidate]; !exists {
			return candidate
		}
	}
}

func reserveBasenames(items []struct {
	Key      string
	Filename string
}, alreadyTaken []string) map[string]string {
	taken := make(map[string]struct{}, len(alreadyTaken)+len(items))
	for _, name := range alreadyTaken {
		if name != "" {
			taken[name] = struct{}{}
		}
	}
	reserved := make(map[string]string, len(items))
	for _, item := range items {
		filename := item.Filename
		if filename == "" {
			filename = item.Key
		}
		name := nextAvailableBasename(filename, taken)
		taken[name] = struct{}{}
		reserved[item.Key] = name
	}
	return reserved
}

File: client/static/js/src/files/output-basename.ts
/**
 * Basename collision helper for multi-file download.
 * photo.png -> photo-1.png -> photo-2.png
 */

export function splitBasenameExtension(filename: string): { stem: string; ext: string } {
  const base = filename.replace(/^.*[/\\]/, '');
  const safe = base.length > 0 ? base : 'download';
  const lastDot = safe.lastIndexOf('.');
  if (lastDot <= 0) {
    return { stem: safe, ext: '' };
  }
  return {
    stem: safe.slice(0, lastDot),
    ext: safe.slice(lastDot),
  };
}

export function nextAvailableBasename(
  desiredName: string,
  taken: ReadonlySet<string>,
): string {
  const { stem, ext } = splitBasenameExtension(desiredName);
  let candidate = `${stem}${ext}`;
  if (!taken.has(candidate)) {
    return candidate;
  }
  let n = 1;
  while (taken.has(`${stem}-${n}${ext}`)) {
    n += 1;
  }
  return `${stem}-${n}${ext}`;
}

/**
 * Reserve unique basenames for a batch. Each target keeps its reserved name
 * across retries -- callers should reuse the returned map rather than
 * re-running reservation after a failed attempt.
 */
export function reserveBasenames(
  items: ReadonlyArray<{ key: string; filename: string }>,
  alreadyTaken: Iterable<string> = [],
): Map<string, string> {
  const taken = new Set<string>();
  for (const name of alreadyTaken) {
    if (name) {
      taken.add(name);
    }
  }
  const reserved = new Map<string, string>();
  for (const item of items) {
    const name = nextAvailableBasename(item.filename || item.key, taken);
    taken.add(name);
    reserved.set(item.key, name);
  }
  return reserved;
}
